File: Dukan_x/my-backend/src/modules/mobile-shop/config/correlation.config.ts
/**
 * Correlation ID Configuration
 *
 * Defines format, propagation rules, and constraints for correlation IDs
 * used in observability, tracing, and audit across the MobileShop system.
 *
 * Requirements: 6.23 (observability), 12.3 (correlation propagation)
 */

export interface CorrelationConfig {
  /** Header name used to propagate correlation ID */
  readonly headerName: string;
  /** Prefix for generated correlation IDs */
  readonly prefix: string;
  /** Maximum length of a correlation ID */
  readonly maxLength: number;
  /** Regex pattern for validation */
  readonly pattern: string;
  /** Whether to generate a new ID if not provided by client */
  readonly generateIfMissing: boolean;
  /** Fields included in telemetry emission */
  readonly telemetryFields: readonly string[];
}

export const CORRELATION_CONFIG: CorrelationConfig = {
  headerName: 'X-Correlation-Id',
  prefix: 'ms-',
  maxLength: 64,
  pattern: '^[a-zA-Z0-9\\-_]{1,64}$',
  generateIfMissing: true,

  telemetryFields: [
    'correlationId',
    'tenantId',
    'operationId',
    'entityType',
    'accessPatternId',
    'tableOrIndex',
    'operationType',
    'consistencyMode',
    'itemCount',
    'consumedCapacity',
    'latencyMs',
    'conditionalResult',
    'transactionResult',
    'retryCount',
    'backoffMs',
    'throttlingReason',
    'throttlingResource',
    'reconciliationOutcome',
  ],
} as const;

File: Dukan_x/lib/features/mobile_shop/config/correlation_config.dart
/// Correlation ID Configuration (Flutter)
///
/// Defines format, header name, and propagation rules for correlation IDs
/// used across API calls, telemetry, and audit trails.
///
/// Requirements: 6.23, 12.3
library;

import 'package:flutter/foundation.dart';

/// Correlation ID configuration.
@immutable
class CorrelationConfig {
  /// HTTP header name used to propagate correlation ID.
  final String headerName;

  /// Prefix for client-generated correlation IDs.
  final String prefix;

  /// Maximum length of a correlation ID.
  final int maxLength;

  /// Regex pattern for validation.
  final String pattern;

  /// Whether to generate a new ID if not provided.
  final bool generateIfMissing;

  const CorrelationConfig({
    required this.headerName,
    required this.prefix,
    required this.maxLength,
    required this.pattern,
    required this.generateIfMissing,
  });

  /// Validates a correlation ID against the configured pattern.
  bool isValid(String id) {
    if (id.isEmpty || id.length > maxLength) return false;
    return RegExp(pattern).hasMatch(id);
  }
}

/// Default correlation configuration.
const kCorrelationConfig = CorrelationConfig(
  headerName: 'X-Correlation-Id',
  prefix: 'ms-',
  maxLength: 64,
  pattern: r'^[a-zA-Z0-9\-_]{1,64}$',
  generateIfMissing: true,
);
